roachtest: pgjdbc failed

roachtest.pgjdbc [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4325206&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4325206&tab=artifacts#/pgjdbc) on master @ [a7b1c17c240c935056f87f29dfc1ee75611c371e](https://github.com/cockroachdb/cockroach/commits/a7b1c17c240c935056f87f29dfc1ee75611c371e):


```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /artifacts/pgjdbc/run_1
	orm_helpers.go:191,orm_helpers.go:117,java_helpers.go:220,pgjdbc.go:188,pgjdbc.go:200,test_runner.go:779: 
		Tests run on Cockroach v22.1.0-alpha.1-975-ga7b1c17c24
		Tests run against pgjdbc REL42.2.19
		5298 Total Tests Run
		4555 tests passed
		743 tests failed
		60 tests skipped
		185 tests ignored
		4 tests passed unexpectedly
		0 tests failed unexpectedly
		0 tests expected failed but skipped
		0 tests expected failed but not run
		---
		--- PASS: org.postgresql.test.jdbc2.BatchedInsertReWriteEnabledTest.test32000Binds[2: autoCommit=NO, binary=REGULAR] - https://github.com/cockroachdb/cockroach/issues/26508 (unexpected)
		--- PASS: org.postgresql.test.jdbc2.BatchedInsertReWriteEnabledTest.testBatchWithReWrittenBatchStatementWithFixedParameter[3: autoCommit=NO, binary=FORCE] - https://github.com/cockroachdb/cockroach/issues/26508 (unexpected)
		--- PASS: org.postgresql.test.jdbc2.BatchedInsertReWriteEnabledTest.test32000Binds[3: autoCommit=NO, binary=FORCE] - https://github.com/cockroachdb/cockroach/issues/26508 (unexpected)
		--- PASS: org.postgresql.test.jdbc2.BatchedInsertReWriteEnabledTest.testBatchWithReWrittenBatchStatementWithFixedParameter[2: autoCommit=NO, binary=REGULAR] - https://github.com/cockroachdb/cockroach/issues/26508 (unexpected)
		For a full summary look at the pgjdbc artifacts 
		
		An updated blocklist (pgjdbcBlockList22_1) is available in the artifacts' pgjdbc log
```
<details><summary>Help</summary>
<p>


See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)



See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)

</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #75649 roachtest: pgjdbc failed [C-test-failure O-roachtest O-robot branch-release-21.1]
- #75547 roachtest: pgjdbc failed [C-test-failure O-roachtest O-robot branch-release-21.2]
</p>
</details>
/cc @cockroachdb/sql-experience
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*pgjdbc.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

roachtest.pgjdbc [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4360904&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4360904&tab=artifacts#/pgjdbc) on master @ [80f1c2ce09389f1d7e97376964d3f2a922405b1b](https://github.com/cockroachdb/cockroach/commits/80f1c2ce09389f1d7e97376964d3f2a922405b1b):


```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /artifacts/pgjdbc/run_1
	orm_helpers.go:191,orm_helpers.go:117,java_helpers.go:220,pgjdbc.go:188,pgjdbc.go:200,test_runner.go:779: 
		Tests run on Cockroach v22.1.0-alpha.1-1182-g80f1c2ce09
		Tests run against pgjdbc REL42.2.19
		5298 Total Tests Run
		4553 tests passed
		745 tests failed
		60 tests skipped
		185 tests ignored
		4 tests passed unexpectedly
		2 tests failed unexpectedly
		0 tests expected failed but skipped
		0 tests expected failed but not run
		---
		--- FAIL: org.postgresql.test.jdbc4.ConnectionValidTimeoutTest.testIsValidRespectsSmallerTimeout[networkTimeoutMillis=1,500, validationTimeoutSeconds=1, expectedMaxValidationTimeMillis=1,100] - unknown (unexpected)
		--- PASS: org.postgresql.test.jdbc2.BatchedInsertReWriteEnabledTest.testBatchWithReWrittenBatchStatementWithFixedParameter[3: autoCommit=NO, binary=FORCE] - https://github.com/cockroachdb/cockroach/issues/26508 (unexpected)
		--- FAIL: org.postgresql.test.jdbc4.ConnectionValidTimeoutTest.testIsValidRespectsSmallerTimeout[networkTimeoutMillis=0, validationTimeoutSeconds=1, expectedMaxValidationTimeMillis=1,100] - unknown (unexpected)
		--- PASS: org.postgresql.test.jdbc2.BatchedInsertReWriteEnabledTest.test32000Binds[2: autoCommit=NO, binary=REGULAR] - https://github.com/cockroachdb/cockroach/issues/26508 (unexpected)
		--- PASS: org.postgresql.test.jdbc2.BatchedInsertReWriteEnabledTest.testBatchWithReWrittenBatchStatementWithFixedParameter[2: autoCommit=NO, binary=REGULAR] - https://github.com/cockroachdb/cockroach/issues/26508 (unexpected)
		--- PASS: org.postgresql.test.jdbc2.BatchedInsertReWriteEnabledTest.test32000Binds[3: autoCommit=NO, binary=FORCE] - https://github.com/cockroachdb/cockroach/issues/26508 (unexpected)
		For a full summary look at the pgjdbc artifacts 
		
		An updated blocklist (pgjdbcBlockList22_1) is available in the artifacts' pgjdbc log
```
<details><summary>Help</summary>
<p>


See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)



See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)

</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #75649 roachtest: pgjdbc failed [C-test-failure O-roachtest O-robot branch-release-21.1]
- #75547 roachtest: pgjdbc failed [C-test-failure O-roachtest O-robot branch-release-21.2]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*pgjdbc.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/sql/pgwire/server.go
+++ b/pkg/sql/pgwire/server.go
@@ -747,16 +747,15 @@ func (s *Server) handleCancel(ctx context.Context, conn net.Conn, buf *pgwirebas
 			log.Sessions.Warningf(ctx, "unexpected while handling pgwire cancellation request: %v", err)
 		}
 		telemetry.Inc(sqltelemetry.CancelRequestCounter)
-
-		// The connection that issued the cancel is not a SQL session -- it's an
-		// entirely new connection that's created just to send the cancel. There
-		// isn't anything left for the server to do after it handles the
-		// pgwire cancel request.
-		_ = conn.Close()
 	}()
 
 	var backendKeyDataBits uint64
 	backendKeyDataBits, err = buf.GetUint64()
+	// The connection that issued the cancel is not a SQL session -- it's an
+	// entirely new connection that's created just to send the cancel. We close
+	// the connection as soon as possible after reading the data, since there
+	// is nothing to send back to the client.
+	_ = conn.Close()
 	if err != nil {
 		return
 	}

